src/js/trove/multiple-regression.js

```diff
@@ -0,0 +1,182 @@
+({
+  requires: [],
+  nativeRequires: ["pyret-base/js/namespace"],
+  provides: {
+    values: {
+      "multiple-regression": "tany"
+    }
+  },
+  theModule: function(runtime, namespace, uri) {
+
+    function makeColumnMatrix(y_s) {
+      let n = y_s.length;
+      let M = new Array(n);
+      for (let i = 0; i < n; i++) {
+        M[i][0] = y_s[i];
+      }
+      return M;
+    }
+
+    function matrixTranspose(m) {
+      let numrows = m.length;
+      let numcols = m[0].length;
+      let mT = new Array(numcols);
+      for (let i = 0; i < numcols; i++) {
+        mT[i] = new Array(m);
+        for (let j = 0; j < numrows; j++) {
+          mT[i][j] = m[j][i];
+        }
+      }
+      return mT;
+    }
+
+    function matrixDeterminant(m) {
+      let size = m.length;
+      if (size === 0) {
+        return 0;
+      }
+      if (m[0].length !== size) {
+        return 0;
+      }
+      if (size === 1) {
+        return m[0][0];
+      }
+      if (size === 2) {
+        return (m[0][0] * m[1][1] - m[1][0] * m[0][1]);
+      }
+      let det = 0;
+      for (let i = 0; i < size; i++) {
+        det += m[0][i] * matrixCofactor(m, 0, i);
+      }
+      return det;
+    }
+
+    function matrixCofactor(m, i, j) {
+      let size = m.length;
+      if (size === 0) {
+        return 0;
+      }
+      if (m[0].length !== size) {
+        return 0;
+      }
+      let mm = new Array(size - 1);
+      for (let ii = 0; ii < size; ii++) {
+        if (ii === i) { continue; }
+        let iii = (ii > i ? ii - 1 : ii)
+        mm[iii] = new Array(size - 1);
+        for (let jj = 0; jj < size; jj++) {
+          if (jj === j) { continue; }
+          let jjj = (jj > j ? jj - 1 : jj)
+          mm[iii][jjj] = m[ii][jj];
+        }
+      }
+      let sign = (i + j) % 2;
+      sign = (sign === 0) ? 1 : -1;
+      return sign * matrixDeterminant(mm)
+    }
+
+    function matrixAdjoint(m) {
+      let size = m.length;
+      let mc = new Array(size);
+      for (let i = 0; i < size; i++) {
+        mc[i] = new Array(size);
+        for (let j = 0; j < size; j++) {
+          mc[i][j] = matrixCofactor(m, i, j);
```

Comment:
Would it not be more efficient to construct the transpose as you go, instead of needing to allocate a whole extra matrix?